proposals/0672-tuist-react-native.md

```diff
@@ -0,0 +1,85 @@
+---
+title: Tuist.io in React Native
+author:
+- Riccardo Cipolleschi
+date: 2023-06-19
+---
+
+# RFC0672: Tuist.io in React Native 
+
+## Summary
+
+Managing Xcodeproj files is error prone and time consuming. 
+
+The current proposal suggests to move away from manually managing `Xcodeproj` files in favor of a declarative description leveraging [Tuist.io](https://tuist.io) as tool.
+
+## Basic example
+
+You can have a look at it in action in this PR:
+
+- https://github.com/facebook/react-native/pull/37952
+
+The idea is to describe the project with a declarative approach using Swift. Tuist allows you to load the `Project.swift` file in an instance of Xcode, leveraging also code completion and type checking to ensure the correctness of the project.
+
+## Motivation
+
+Managing Xcodeproj files is error prone and time consuming. Those files are source of frustrations for teams using them and they have been reported as one of the main pain point when users have to migrate to a new version of React Native.
+
+They are also an old way to manage projects. Even Apple realized that they are not the best approach and started developing Swift Package Manager (SwiftPM) that replace the project for developing libraries.
+
+Moder approaches relies on declarative syntax to describe how the project should look like and tools that can generate a valid Xcodeproj starting from that description.
+
+There are several tools that we can use, but the proposed one is [Tuist.io](https://tuist.io).
+
+[Tuist.io](https://tuist.io) is OSS and currently maintained by several companies, among which Shopify, which is already a React Native partner.
+
+It allows the users to describe their Apple projects in Swift (the default language for iOS), and it also leverage Xcode for type checking and code completion, ensuring that the project definition is valid at compile time.
+
+## Detailed design
+
+To adopt Tuist in React Native, we need to follow some steps:
+
+1. Land https://github.com/facebook/react-native/pull/37952. This PR already contains a functioning project which passed all the iOS tests in CI.
+2. Update the CLI to install Tuist in the user machine. Tuist comes directly with a `tuistenv`additional tool, which handle the environment and version for the user.
```

Comment:
Have you considered an approach similar to the one that tools like ESBuild follow where the binary is downloaded when the NPM packages is installed using hooks? It'd require us (Tuist) to maintain the NPM package, but they'd be able to pin the version of Tuist using the package manager's pinning solution, and also cache Tuist with the caching of all the `node_modules` in environments like CI.
Just putting it out there for further discussion.

Reply:
Hi @pepicrft, thanks for the suggestion.

I am not very familiar with the NPM hooks and workflows. If Tuist is open to maintain an NPM package, surely it would be an improvement over the current proposal.

Reply:
Resolving this, added the proposed idea in the RFC